### ISSUE TITLE
[clang] Reporting unused function with ifunc resolver

### ISSUE THREAD

**Issue by JiaT75** [llvm/llvm-project#63957]
Hello!

Clang 16.0.0 and older report a warning when compiling a program that uses `__attribute__((__ifunc__()))` with `-Wunused-function`. The resolver function appears to trigger this when it is only referenced by the `__attribute__`. Here is an example program that should trigger the warning:

```C
static void resolved(void) {
    return;
}

static void (*resolver (void)) (void) {
	return resolved;
}

void indirect_function(void) __attribute__((__ifunc__("resolver")));

int main() {
    indirect_function();
}
```
GCC does not issue a warning for the above program. [Here](https://godbolt.org/z/9f1TnjG7h) is a godbolt.org snippet of the above code to allow you to easily test it. Let me know if you need anymore information.

**Comment by Zentrik** [comment]
This is still an issue on trunk.

**Comment by mxmlnkn** [comment]
Should this really be changed taking into account the ifunc-based xz backdoor added by the issue reporter? https://www.openwall.com/lists/oss-security/2024/03/29/4

**Comment by tstellar** [comment]
We should not make this change  in LLVM right now.

**Comment by MaskRay** [comment]
> We should not make this change in LLVM right now.

https://github.com/tukaani-project/xz/commit/82ecc538193b380a21622aea02b0ba078e7ade92 is a legitimate change that exposes two problems:

* Static linking with glibc and gcc -fprofile-generate can lead to crashes. The instrumented resolver attempts to access TLS before it's properly set up by __libc_setup_tls (https://gcc.gnu.org/PR114115). https://maskray.me/blog/2021-01-18-gnu-indirect-function#non-preemptible-ifunc#:~:text=no_profile_instrument_function
* This Clang -Wunused-function false positive, which should be mostly fixed by #87130

We should be careful and don't consider all xz commits malicious.

**Comment by tstellar** [comment]
> We should be careful and don't consider all xz commits malicious.

I disagree, we need to be extra careful with changes associated with the xz project right now.  There's no reason why we need to get this change committed quickly.  It should get extra scrutiny.

**Comment by shafik** [comment]
> > We should be careful and don't consider all xz commits malicious.
> 
> I disagree, we need to be extra careful with changes associated with the xz project right now. There's no reason why we need to get this change committed quickly. It should get extra scrutiny.

I concur with this sentiment, the details are still emerging on this whole incident. This so far seems like the best timeline: https://boehs.org/node/everything-i-know-about-the-xz-backdoor and a lot if still unknown.

I don't feel confident in saying I have the whole picture and I think until we get a better idea of the scope we should be taking our time.

**Comment by NuLL3rr0r** [comment]
Did GitHub just restored their access? I thought the account was suspended.

**Comment by vvdaal** [comment]
Rather offtopic but 
> Did GitHub just restored their access? I thought the account was suspended.

No (check https://github.com/Larhzu?tab=following), Larhzu is also still suspended. (https://github.com/JiaT75?tab=following)

**Comment by ItzSwirlz** [comment]
literally seeing this pr, if Jia could read this I'd say:

![image](https://github.com/llvm/llvm-project/assets/24401303/9971aef5-3829-4cf7-aa1f-d787558c5c50)

**Comment by thesamesam** [comment]
I'm going to lock this for now but we can unlock it later when things have quietened down.
